### PR TITLE
[v2int/1.1] Restore bindToContext functionality

### PR DESCRIPTION
**Pull request by markfields**
After discussing the challenges being faced by our internal partners integrating the 2.0.0-internal.1.1.x release due to the removal of `bindToContext` (see #10627), we're adding it back in.  They are tracking the work required to fully migrate off the request pattern to using handles, which will render this fully obsolete and at that point we will remove it.

## Does this introduce a breaking change?

No... it undoes a breaking change, such that this change is additive.  Typically I think we avoid even backwards-compatible API changes in patch releases, but all interfaces being modified are for internal use only (we don't expect partners to be implementing them).  And this is required to unblock our partner integrating the release.

### CODE REVIEW

File: api-report/container-runtime-definitions.api.md
Comment (anthony-murphy) [inline api-report/container-runtime-definitions.api.md:72]: is there a need to extend data store here?

File: api-report/runtime-definitions.api.md
Comment (anthony-murphy) [inline api-report/runtime-definitions.api.md:168]: maybe make this optional, that will make it non-breaking if people don't care about his api

File: packages/test/test-end-to-end-tests/src/test/deRehydrateContainerTests.spec.ts
Comment (anthony-murphy) [inline packages/test/test-end-to-end-tests/src/test/deRehydrateContainerTests.spec.ts:568]: these changes don't look necessary

**Comment by anthony-murphy** [comment]
looks ok. in general, i would usually check in in main first, then back port to all necessary branches

File: packages/test/test-end-to-end-tests/src/test/deRehydrateContainerTests.spec.ts
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/deRehydrateContainerTests.spec.ts:568]: Ironic, huh.  It was part of the revert.  But I'll skip it for this branch.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreAliased.spec.ts
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreAliased.spec.ts:19]: It should now pe possible to run full compat tests and will give better coverage.

**Comment by agarwal-navin** [comment]
Looks good. Can you please add tests for the scenario where in detached container, data stores are requested while its parent data store is initializing.

**Review by agarwal-navin — APPROVED** [review]

File: api-report/container-runtime-definitions.api.md
Comment (markfields) [reply api-report/container-runtime-definitions.api.md:72]: Yeah I think it's correct this way.  The thing being cast to this is IDataStore, and Bohemia uses it as such (not just to call bindToContext).  https://dev.azure.com/office/OC/_git/office-bohemia?path=/packages/base-container/src/handlers/createComponentHandler.ts&version=GBmaster&line=316&lineEnd=330&lineStartColumn=1&lineEndColumn=4&lineStyle=plain&_a=contents

File: api-report/runtime-definitions.api.md
Comment (markfields) [reply api-report/runtime-definitions.api.md:168]: I may just leave it off here altogether and have FluidDataStoreRuntime.bindToContext call this.dataStoreContext.makeLocallyVisible() directly rather than having the bindToContext shim on DataStoreContext.  Will make this PR smaller, but less of a direct revert.

File: packages/test/test-end-to-end-tests/src/test/bindToContextTests.spec.ts
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/bindToContextTests.spec.ts:58]: Remove console logs.

File: packages/test/test-end-to-end-tests/src/test/bindToContextTests.spec.ts
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/bindToContextTests.spec.ts:61]: Basically, what you are doing in the first test.

File: api-report/runtime-definitions.api.md
Comment (markfields) [reply api-report/runtime-definitions.api.md:168]: Leaving it on for now, it really needs to be there for the code I'm adding back to work.  I'll refactor it away in next once these changes get there.

**Comment by markfields** [comment]
Will merge after #12062 is merged into main (same change)

**Comment by markfields** [comment]
Update: Planning to hold on this until next week when FF/Bohemia stakeholders have a chance to discuss and agree upon a long term plan for this API (i.e. when we can remove it for good)

**Pull request merged**
